### DISCUSSION TITLE
Update attributes in `add_node` explicitly by a flag.

### DISCUSSION THREAD

**Discussion by 4l1fe** [networkx/networkx#6025] [General]
Hey, i have a case when i'm adding nodes which might exist already with specific attributes and when i'm encountering it again those attributes are overwritten.

I'm doing something like this:
```python
for node in nodes:
    graph.add_node(node, **attrs)  # a node is added here with `attrs`

    for subnode in form_subnodes(node):
        graph.add_node(subnode, **attrs) # the node above can be contained in the result returned by function and `attrs` are overwritten.
```

Would be good to have a flag like `update_attrs=False` in `node_add()`. What do you think?

```python
for node in nodes:
    graph.add_node(node, **attrs)

    for subnode in form_subnodes(node):
        graph.add_node(subnode, **attrs, update_attrs=False)  # now i'm sure `attrs` aren't overwritten.
```

[Here is code](https://github.com/networkx/networkx/blob/main/networkx/classes/graph.py#L560) which do it in `add_node`:
```python
        if node_for_adding not in self._node:
            if node_for_adding is None:
                raise ValueError("None cannot be a node")
            self._adj[node_for_adding] = self.adjlist_inner_dict_factory()
            attr_dict = self._node[node_for_adding] = self.node_attr_dict_factory()
            attr_dict.update(attr)
        else:  # update attr even if node already exists
            self._node[node_for_adding].update(attr)
```

for the time being, i get a side effect as i don't need the attributes to get updated.

**Comment by dschult** [comment]
Have you tried:
```python
for node in nodes:
    graph.add_node(node, **attrs)  # a node is added here with `attrs`

    for subnode in form_subnodes(node):
        if subnode not in graph:
            graph.add_node(subnode, **attrs)
```

**Reply by 4l1fe** [reply]
> Have you tried:
> 
> ```python
> for node in nodes:
>     graph.add_node(node, **attrs)  # a node is added here with `attrs`
> 
>     for subnode in form_subnodes(node):
>         if subnode not in graph:
>             graph.add_node(subnode, **attrs)
> ```

Sure i can code so, but this check is a [duplication](https://github.com/networkx/networkx/blob/main/networkx/classes/graph.py#L554) of
```python
    def add_node(self, node_for_adding, **attr):
        ....
        if node_for_adding not in self._node:
            ...
```

What are your thoughts on the update flag?
I wonder if you don't find updating attributes an unexpected effect when the method is about adding only a node in regard to its name `add_node`? It's similar to the DB `INSERT, UPDATE` operations semantic in terms of an operation has the only effect.

**Comment by dschult** [comment]
The function is intended to add nodes if necessary and update the attributes. I don't see updating the attributes as an unintended effect. Whatever the last attributes given to the graph are the attributes in the graph. 

There are so many potential optional keyword arguments for so many functions. Our design attempts to avoid such arguments where possible. And I think it is potentially confusing: if someone doesn't want the attributes updated, why are they supplying the attributes? 

The addition of the time to lookup a node is very small -- even compared to a function call. 

With the caveat that timing python code can be very counterintuitive sometimes, so this might not be slower at all, here's another possible code option that is presumably faster if speed is a concern:
```python
nodes_to_add = ((node, attrs) for n in form_subnodes(node) if n not in graph)
graph.add_nodes_from(nodes_to_add)
```

**Reply by 4l1fe** [reply]
>  And I think it is potentially confusing: if someone doesn't want the attributes updated, why are they supplying the attributes?

Well, my point is that `add_node` doesn't mean `update_attributes` and as there is the check inside `add_node` for an existing node i ask myself why logic that is out of the method name context takes  place, it wasn't declared...  

I encountered many cases working with data when an operation has to be atomic, either `insert` or `update`. Sometimes an operation consist of both effects and is called like `upsert` and you expect data to be inserted if not exist or updated otherwise. 

In my specific case in the loop above, as a node was already added with the attributes, i need the `add_node` method to skip logic and do nothing, this is already implemented inside [the first condition](https://github.com/networkx/networkx/blob/main/networkx/classes/graph.py#L554), but the second condition is ran which adds the second effect. It looks like an operation `upsert` and doesn't fit all the cases. The flag `update_attrs` could enable or disable the second effect according to my need as there no **atomic method** to add nodes.

> There are so many potential optional keyword arguments for so many functions

I understand.

> The addition of the time to lookup a node is very small -- even compared to a function call.

I agree, speed isn't a concern.

Anyway, i've heard you and am going to work it around myself. Thank you for sharing your view.
